teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

```diff
@@ -0,0 +1,88 @@
+---
+status: proposed
+title: Embedded TaskRuns and Runs Status in PipelineRuns
+creation-date: '2022-01-24'
+last-updated: '2022-01-24'
+authors:
+- '@lbernick'
+- '@jerop'
+---
+
+# TEP-0100: Embedded TaskRuns and Runs Status in PipelineRuns
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Requirements](#requirements)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes changes to `PipelineRun` Status to reduce the amount of
+information stored about the status of its `Run`s and `TaskRun`s to improve
+performance, reduce memory bloat and improve extensibility.
+
+## Motivation
+
+Today, we embed the status of each `TaskRun` and `Run` into the `PipelineRun`
+status. This causes several issues:
+
+1. **Performance degradation**. Every time a `TaskRun`'s or `Run`'s status changes,
+   the parent `PipelineRun`'s status is updated as well. For example, the `PipelineRun`'s
+   status is updated upon completion of each `Step`, even if the `TaskRun` or `Run`
+   has not completed. This causes extra requests to etcd and extra load on the Dashboard,
+   which reacts to CRD status updates. Read more in the [related issue][issue-3140].
+
+2. **Memory bloat**. Embedded statuses increase the size of the serialized `PipelineRuns`.
+   As shared in [API WG on 10/01/2022][api-wg], the embedded statuses:
+   > "more than doubles the storage and has direct consequences on what customers end up paying"
+
+3. **Lack of extensibility**. The above problems will be exacerbated by features that will execute
+   multiple `TaskRuns` and `Runs` from one `PipelineTask`. For example:
+   * [**`Matrix`**][tep-0090]: fan out a given `PipelineTask` into multiple `TaskRuns` or `Run`s.
+     Matrix `Run`s can even be created dynamically by consuming `Results` of previous `Run`s
+   * [**`Pipelines` in `Pipelines`**][tep-0056]: pass in `Pipelines` to `PipelineTasks` so we can
+     run them similarly to `Tasks`.
+
+### Goals
+
+- Reduce updates to `PipelineRun` status and the amount of storage it uses.
+- Set up `PipelineRun` status to better support future extensions in Tekton Pipelines.
+
+### Non-Goals
+
+- Improving other aspects of `PipelineRun` status other than the embedding of `TaskRuns` and
+  `Runs`.
+
+### Requirements
+
+`PipelineRun` status must:
+- show the`TaskRuns` or `Runs` that were executed.
+- show the execution status of each `TaskRun` (e.g. Successful, Failed).
+- support future extensions that would create even more `TaskRuns` and `Runs`.
+- optimize for performance and storage.
```

Comment:
pipelineRun status contains `startTime`, `completionTime`, `conditions` (indicating success or failure of pipelineRun itself), `pipelineResults`, `pipelineSpec`,  a map of `taskRuns`, and a map of `runs`.

Here, we are trying to optimize the map of `taskRun` and `run`. Each `taskRun` and `run` in turn contains `startTime`, `completionTime`, `conditions` (indicating success or failure of taskRun/run itself), `podName`, a list of steps, `taskResults`, and `taskSpec`.

Reply:
it may be nice to mention what are we "loosing" out of the optimizations, which fields are going to be not available anymore for tools/users to parse.

an example of the expected status would be nice too,

Reply:
Thank you for the reviews @pritidesai and @chmouel!

@pritidesai added a background section providing more information about the `PipelineRunStatus` as you described above

@chmouel we plan to discuss what fields will be removed and provide examples in the proposal (which will be added in the next pull request) - added an "open questions" section with your question above to ensure they will be addressed in the proposal